Divide the circle : find the indices, number of pixels, pixel indices

sameera2004 opened the pull request.

tacaswell commented on skxray/diff_roi_choice.py:
```diff
@@ -375,3 +378,117 @@ def _process_rings(num_rings, img_dim, ring_vals, ring_inds):
+    return roi_inds, num_pixels, pixel_list
+
+
+def _process_rois(mesh, detector_size, num_rois):
```
wasn't there already a method with this name?

sameera2004 replied:
Yes _process_rings() changed it, now it will only process rings edge values. 
This module will take care of all the modules final processing of finding pixel_list, number of pixels, indices

tacaswell commented on skxray/tests/test_diff_roi_choice.py:
```diff
@@ -185,3 +185,41 @@ def _helper_check(pixel_list, inds, num_pix, q_ring_val, calib_center,
+
+    num_pixels_m = (np.bincount(ty.astype(int)))[1:]
+
+    return (ty.reshape(*detector_size))
```
Tests should not be returning anything

sameera2004 replied:
Yes, this PR is in Progress

tacaswell commented on skxray/tests/test_diff_roi_choice.py:
```diff
@@ -185,3 +185,41 @@ def _helper_check(pixel_list, inds, num_pix, q_ring_val, calib_center,
+
+
+if __name__=="__main__":
+    import matplotlib.pyplot as plt
```
and the main on this should run nose on the file

sameera2004 replied:
Sure, I use matplotlib to see, whether code is working

sameera2004 commented on skxray/tests/test_diff_roi_choice.py:
```diff
@@ -185,3 +187,60 @@ def _helper_check(pixel_list, inds, num_pix, q_ring_val, calib_center,
+                           6])
+
+    assert_array_equal(num_pixels_m, num_pixels)
+    assert_array_equal(roi_inds, roi_inds_m)
```
![divide_ring](https://cloud.githubusercontent.com/assets/5701652/6448963/78d145e2-c0eb-11e4-8069-507478014f1a.png)

sameera2004 replied:
@tacaswell  now this PR is ready for review

tacaswell replied:
What code generated this?

tacaswell commented on skxray/diff_roi_choice.py:
```diff
@@ -157,12 +160,12 @@ def roi_rings(img_dim, calibrated_center, num_rings,
     ring_inds : ndarray
         indices of the required rings
 
+    ring_vals : ndarray
```
Good catch

tacaswell commented on skxray/diff_roi_choice.py:
```diff
@@ -319,28 +322,52 @@ def _grid_values(img_dim, calibrated_center):
-    This will find the indices of the required rings, find the bin
-    edges of the rings, and count the number of pixels in each ring,
-    and pixels list for the required rings.
+    This will process the edge values of the each roi ring
```
Can you explain what sort of processing will be done by this function?

sameera2004 replied:
Added an explanation for _process_rings(num_rings, ring_vals):

tacaswell commented on skxray/diff_roi_choice.py:
```diff
@@ -319,28 +322,52 @@ def _grid_values(img_dim, calibrated_center):
-        indices of the required rings
-        shape is ([detector_size[0]*detector_size[1]], )
+    """
+    ring_vals = np.array(ring_vals)
```
This can be done in one line as `np.asarray(ring_vas).reshape(-1, 2)`

tacaswell commented on skxray/diff_roi_choice.py:
```diff
@@ -360,18 +383,83 @@ def _process_rings(num_rings, img_dim, ring_vals, ring_inds):
+
+    Parameters
+    ----------
+    mesh : ndarry
```
What is 'mesh'?

tacaswell commented on skxray/diff_roi_choice.py:
```diff
@@ -291,9 +293,10 @@ def roi_rings_step(img_dim, calibrated_center, num_rings,
-                                  ring_vals, ring_inds)
+    ring_vals = _process_rings(num_rings, np.asarray(ring_vals))
+
+    ring_inds, num_pixels, pixel_list = _process_rois(np.ravel(ring_inds),
```
This looks like you are un-raveling it so that you can then ravel it.

tacaswell commented on skxray/diff_roi_choice.py:
```diff
@@ -360,18 +383,83 @@ def _process_rings(num_rings, img_dim, ring_vals, ring_inds):
-    grid = np.indices((img_dim[0], img_dim[1]))
-    pixel_list = (grid[0]*img_dim[1] + grid[1]).flatten()[w]
+    mesh = np.zeros(detector_size, dtype=np.int64)
+    angles = np.linspace(0, 360, num_angles)
```
Can we work in radians internally?

tacaswell commented on skxray/diff_roi_choice.py:
```diff
@@ -360,18 +383,83 @@ def _process_rings(num_rings, img_dim, ring_vals, ring_inds):
-    ring_vals = ring_vals.reshape(num_rings, 2)
+    angle_grid[grid_values > radius] = np.nan
+
+    for i in range(num_angles-1):
```
I think if you do this an `np.floor(angle_grid / (2*np.pi / num_angles)))` you can then use `bin_count` again.

tacaswell replied:
and I would do the angle filtering _first_, then the radius filtering.

sameera2004 replied:
Ok now I am thinking that will be better too,(angle filtering first, then the radius filtering)

tacaswell commented on skxray/diff_roi_choice.py:
```diff
@@ -319,28 +322,52 @@ def _grid_values(img_dim, calibrated_center):
+    return ring_vals
+
+
+def roi_divide_circle(detector_size, radius, calibrated_center,
```
Should this also take a `start_angle` input?

sameera2004 replied:
No here I am using number of angles that ring(360) divided into

tacaswell replied:
but you can rotate where '0' is relative to the y-axis.

sameera2004 commented:
Made changes according to comments. @tacaswell ready for review

licode commented on skxray/tests/test_diff_roi_choice.py:
```diff
@@ -185,3 +187,60 @@ def _helper_check(pixel_list, inds, num_pix, q_ring_val, calib_center,
+    y_ = (np.flipud(yy) - calibrated_center[1])
+    x_ = (xx - calibrated_center[0])
+    grid_values = np.float_(np.hypot(x_, y_))
+    angle_grid = np.arctan2(y_, x_)*180/(np.pi)
```
use np.rad2deg

licode commented on skxray/diff_roi_choice.py:
```diff
@@ -360,18 +383,85 @@ def _process_rings(num_rings, img_dim, ring_vals, ring_inds):
+        ([detector_size[0]*detector_size[1]], )
+
+    num_pixels : ndarray
+        number of pixels in each roi's
```
This information can be calculated from roi_inds. Do you still want to return this again?

celiafish commented on skxray/tests/test_diff_roi_choice.py:
```diff
@@ -185,3 +187,60 @@ def _helper_check(pixel_list, inds, num_pix, q_ring_val, calib_center,
+
+
+def test_roi_divide_circle():
+    num_rois = 3
```
Please add description of this function

sameera2004 commented:
![rotate_the_indices_roi_circle_divide](https://cloud.githubusercontent.com/assets/5701652/6738808/cc3961f8-ce4a-11e4-8c78-48106934fc3b.png)

to make angles measured from vertical-anti clock wise

sameera2004 commented:
Thanks for the comments. Made changes , now the angle can start measuring from vertical direction, added documentation to roi_rectangels, made some changes to tests, added a detail commit, tests passing. Ready for PR review.

tacaswell commented on skxray/diff_roi_choice.py:
```diff
@@ -56,14 +59,17 @@
 
 
-def roi_rectangles(num_rois, roi_data, detector_size):
+def roi_rectangles(num_rois, roi_data, detector_size, mask=None):
```
mask is not used any where and is not documented.

sameera2004 replied:
Sure, I will take out the mask for now, later I need to add it.

tacaswell commented on skxray/diff_roi_choice.py:
```diff
@@ -56,14 +59,17 @@
-def roi_rectangles(num_rois, roi_data, detector_size):
+def roi_rectangles(num_rois, roi_data, detector_size, mask=None):
     """
+    This module will provide pixel indices, number of pixels, indices
```
function not module

tacaswell commented on skxray/diff_roi_choice.py:
```diff
@@ -319,28 +325,55 @@ def _grid_values(img_dim, calibrated_center):
+        number of angles ring divide into
+        angles are measured from horizontal-anti clock wise
+
+    rotate : {'Y', 'N'}, optional
```
Wouldn't this be better to make this an angle and add a `direction` kwarg`  ('cw' or 'ccw').

sameera2004 replied:
If we keep angle as a user input I don't think it will be a good idea. Because I want to divide the circle into same size wedges and later use that different wedges roi information(pixel indices, indices..) to run one time correlation and see whether there is a difference between those wedges correlation plots.

I kept the rotation as a optional input

tacaswell commented on skxray/diff_roi_choice.py:
```diff
@@ -360,18 +389,83 @@ def _process_rings(num_rings, img_dim, ring_vals, ring_inds):
 
     """
+    #mesh = np.zeros(detector_size, dtype=np.int64)
+    angles = np.linspace(0, 360, num_angles)
```
We really should work in radians internally.

tacaswell commented:
I thought you had changed the API for there to return the labeled arrays?

sameera2004 commented:
I changed that in the one time Correlation, this one has to rebase once that code merged. Therefore, have to review #254 first

sameera2004 commented:
Closed due to merge conflict (Github message : We can’t automatically merge this pull request.)

tacaswell commented:
@sameera2004 You should just re-base on to current master, fix the merge confilcts, and then force-push back to this branch.
